### PR TITLE
Improve PositionColumn to allow more route params

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.5.x
| Description?  | Improve PositionColumn you can now insert an array of parameters into the udpate_route (extracted from the grid result). This allow this extension to manage more use cases and the javascript code is simpler since it doesn't need to send the extra parameters like parentId (which are now in the url) and it focuses on what it needs to do, send the positions.
| Type?         | improvement
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | 
| How to test?  | 

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/11063)

### CODE REVIEW

File: src/PrestaShopBundle/Twig/DataFormatterExtension.php [hunk 1]
```diff
@@ -96,6 +98,39 @@ public function unsetElement($array, $key)
...
+     * @param array $array
+     * @param array $extractedKeys
+     *
+     * @return array
+     */
+    public function extractArray(array $array, array $extractedKeys)
+    {
+        $extractedArray = [];
+        foreach ($extractedKeys as $key => $value) {
+            if (is_int($key)) {
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Twig/DataFormatterExtension.php:119]: No need if / else.
```php
$oldKey = is_int($key) ? $value : $key;
$newKey = $value;
```

Comment (Quetzacoalt91) [reply src/PrestaShopBundle/Twig/DataFormatterExtension.php:119]: Is it actually needed to copy `$value` in `$newKey`?

File: src/PrestaShopBundle/Twig/DataFormatterExtension.php [hunk 2]
```diff
@@ -96,6 +98,39 @@ public function unsetElement($array, $key)
...
+     * ex:
+     *  extractArray(['first_name' => 'John', 'last_name' => 'Doe'], ['first_name']) => ['first_name' => 'John']
+     *  extractArray(['first_name' => 'John', 'last_name' => 'Doe'], ['first_name' => 'name']) => ['name' => 'John']
+     *
+     * @param array $array
+     * @param array $extractedKeys
+     *
+     * @return array
+     */
+    public function extractArray(array $array, array $extractedKeys)
```
Comment (Quetzacoalt91) [inline src/PrestaShopBundle/Twig/DataFormatterExtension.php]: This reminds me of the Laravel Helper `array_pluck`: https://laravel.com/docs/5.7/helpers#method-array-pluck

File: src/PrestaShopBundle/Twig/DataFormatterExtension.php (hunk 1 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Twig/DataFormatterExtension.php:119]: yeah I think the ternary form will work
@Quetzacoalt91 I don't see why the copy is needed since it is inserted in `$extractedArray` straight away

File: src/PrestaShopBundle/Twig/DataFormatterExtension.php (hunk 2 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Twig/DataFormatterExtension.php]: you're right, although the format is different and it only extract one field
do you want me to change the name? I wasn't 100% sure about the naming but I could not find any better one

File: src/PrestaShopBundle/Twig/DataFormatterExtension.php (hunk 1 above)
Comment (Quetzacoalt91) [reply src/PrestaShopBundle/Twig/DataFormatterExtension.php:119]: My mind precisely. :)

Comment (jolelievre) [reply src/PrestaShopBundle/Twig/DataFormatterExtension.php:119]: but on the insertion `$value` is copied inside the array, so even when it changes value during the loop the array content won't change, besides it is not even the value inserted it is only the key used at this iteration

So as @PierreRambaud suggested we can write:

```php
foreach ($extractedKeys as $key => $value) {
      $oldKey = is_int($key) ? $value : $key;
      $newKey = $value;
      if (isset($array[$oldKey])) {
            $extractedArray[$newKey] = $array[$oldKey];
      }
}
```

or even (but I am afraid it "hides" the logic and might be difficult to understand later)

```php
foreach ($extractedKeys as $key => $extractedKey) {
      $oldKey = is_int($key) ? $extractedKey : $key;
      if (isset($array[$oldKey])) {
            $extractedArray[$extractedKey] = $array[$oldKey];
      }
}
```

Comment (tomas862) [reply src/PrestaShopBundle/Twig/DataFormatterExtension.php:119]: Maybe not make thing so complex. Let's assume they both are multidimensional arrays always and are merged by column record.

Also, maybe instead of `extractArrays` we could rename it to `mergeByColumnRecord` or something so it's closer to our use-case.

and the implementation I think can be:

```php
        $records = [
            'first_record' => 1,
            'second_record' => 2,
            'third_record' => 3,
            'fourth_record' => 4,
        ];

        $routes = [
            'first_record' => 'myRouteId',
            'third_record' => 'anotherRouteId',
            'fourth_record' => 'myRouteId',
        ];

        $intersectedRecords = array_intersect_key($records, $routes);

        $result = [];
        foreach ($intersectedRecords as $key => $recordValue) {
            if (!isset($routes[$key])) {
                continue;
            }

            if (!isset($result[$routes[$key]])) {
                $result[$routes[$key]] = $recordValue;
            }
        }
```

Comment (jolelievre) [reply src/PrestaShopBundle/Twig/DataFormatterExtension.php:119]: well talking about complexity it took me a while to understand the whole code ^^
actually there's a reason why I called it `extractArrays` because I don't want this piece of code to be related in any way to the router or url generation, or even a column of record
at first I wanted to a twig function that would look like this:

`build_route(routeName, routeAssociations, result)`

which would automatically extract the data from `result` using the assignations set in `routeAssociations` (poor naming I know, but you get the idea)

Then I realized this `build_route` method would have way too much responsibility, merging, extracting parameters and generating a route, requiring a dependency on the `UrlGenerator` whereas all the parameter conversion could be performed beforehand and simply injected in the already existing `path` or `url` twig helpers

So, long story short ^^, that's what led me to simply build a help which would extract the needed arguments and, if needed, rename them

Comment (tomas862) [reply src/PrestaShopBundle/Twig/DataFormatterExtension.php:119]: all right - then , just maybe the name `extractArrays` is not the best one but again, can't think of better one right now :smile:

Comment (jolelievre) [reply src/PrestaShopBundle/Twig/DataFormatterExtension.php:119]: yep i agree with you maybe `arrayPluck`?
as @Quetzacoalt91 like in Laravel

Comment (jolelievre) [reply src/PrestaShopBundle/Twig/DataFormatterExtension.php:119]: or `stripArray` ? ^^

Comment (mickaelandrieu) [reply src/PrestaShopBundle/Twig/DataFormatterExtension.php:119]: naming aside (I'm ok with `arrayPluck`), I'm ok with this PR :+1:

**Review by mickaelandrieu — DISMISSED** [review]

File: src/PrestaShopBundle/Twig/DataFormatterExtension.php (hunk 1 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Twig/DataFormatterExtension.php:119]: ok I like it as well, done!
if someone can approve this PR we can then merge it and push the module to QA

**Review by mickaelandrieu — DISMISSED** [review]

**Review by mickaelandrieu — APPROVED** [review]

**Comment by mickaelandrieu** [comment]
Thanks @jolelievre !
